### ISSUE TITLE
Question about execution pause

### ISSUE THREAD

**Issue by diamantisk** [rhaiscript/rhai#768]
Is there any way or workaround to pause the execution, store context and resume the execution and context from that point?

**Comment by emesare** [comment]
Yes, I think what you want is the [Debugging Interface](https://rhai.rs/book/engine/debugging/debugger.html) (enabled through the flag `debugging`).

**Comment by schungx** [comment]
> Is there any way or workaround to pause the execution, store context and resume the execution and context from that point?

Not exactly what you like to do.  You have just described _async_, which Rhai is not.

However, as @emesare mentioned, you can use the debugging interface or use the `on_progress` callback.  The pitfalls is that you usually need to use a dedicated thread to stop the execution and another thread to control it.

**Comment by diamantisk** [comment]
Thank you for noticing this, I will experiment with it. 
I mostly imagined a way that when rhai calls a host function defined in rust, to have the engine somehow available in scope and execute a pause, get and store the context, then perform the async task on the rust side and once it completes, rust will restore the context and resume the execution of Rhai script.

**Comment by schungx** [comment]
You can probably do that easily by encapsulating a channel inside `on_progress`.  Use a dedicated thread to run the engine, and use messages to control whether `on_progress` stalls, waiting for the resume command.  There is no need to use the debugging interface if you just want simple start/stop.  There is also no need to store any context, as the thread acts as the context.

**Comment by schungx** [comment]
I have coded up an example: https://github.com/schungx/rhai/blob/master/examples/pause_and_resume.rs

**Comment by schungx** [comment]
> I envision a method where a host function, defined in Rust, is called by Rhai

If you want such a `suspend` function, it is extremely to register one.  Simply `move` in a pair of channels.  When called, send a message to the `tx` channel, and wait for a reply from the `rx` channel.  In the meantime, the thread suspends waiting for the channel message, which is exactly what you want to achieve.
